Source/Classes/PINProgressiveImage.m

```diff
@@ -102,19 +105,56 @@ - (CFTimeInterval)estimatedRemainingTimeThreshold
     return estimatedRemainingTimeThreshold;
 }
 
-- (void)setStartTime:(CFTimeInterval)startTime
+- (float)bytesPerSecond
 {
-    [self.lock lock];
-        _startTime = startTime;
-    [self.lock unlock];
+    __block float bytesPerSecond;
+    [self.lock lockWithBlock:^{
+        bytesPerSecond = [self l_bytesPerSecond];
+    }];
+    return bytesPerSecond;
 }
 
-- (CFTimeInterval)startTime
+- (float)l_bytesPerSecond
 {
-    [self.lock lock];
-        CFTimeInterval startTime = _startTime;
-    [self.lock unlock];
-    return startTime;
+    NSAssert(_dataTask.PIN_startTime != 0, @"Start time needs to be set by now.");
+    CFTimeInterval endTime = CACurrentMediaTime();
+    if (_dataTask.PIN_endTime != 0) {
+        endTime = _dataTask.PIN_endTime;
```

Comment:
The logic here is a little bit hard to follow. Let's do something like `CFTimeInterval endTime = (_dataTask.PIN_endTime ?: CACurrentMediaTime())`